### ISSUE TITLE
Remove gulp-clean-css from blacklist

### ISSUE THREAD

**Issue by scniro** [gulpjs/plugins#210]
package `gulp-clean-css` is blacklisted with the following remarks...

> "gulp-clean-css": "duplicate of gulp-minify-css"

However, `gulp-minify-css` is also blacklisted with the following remarks...

> "gulp-minify-css": "deprecated. use gulp-cssnano."

The author for `gulp-minify-css` has recently deprecated this package in favor of [cssnano](https://www.npmjs.com/package/cssnano) which I strongly feel is not appropriate. I have amassed (while small) a group who also feels this package should not be deprecated. [I have raised and since closed out an issue](https://github.com/murphydanger/gulp-minify-css/issues/137) after taking joint ownership of `gulp-clean-css` on npm, which had previously laid dormant for two years and was published in a quick-fix situation without a long term vision. `gulp-minify-css` appears to be dead - I want to bring it back even better.

I have made progress today, and will continue to make progress. My goal is to provide to go-to gulp plugin for `clean-css`, but it has to start somewhere. Removing this package from the blacklist database would be a great catalyst to get this  going.

**Comment by sindresorhus** [comment]
:+1: `cssnano` is good, but not yet on the level of `clean-css`.

**Comment by phated** [comment]
@scniro can you discuss further what your vision is here?  Maybe even address why you think there should be both cssnano and gulp-minify-css side by side?

**Comment by scniro** [comment]
If you check out the issue I opened up in my initial post, `cssnano` was incredibly slow for a large enterprise application I am working on. Minifying ~500+ fairly large files absolutely choked, with execution times reaching **tens** of **minutes.** I did my best to investigate the issue and drew it back to a [raised cssnano issue](https://github.com/ben-eb/cssnano/issues/47) which is my best documented reasoning to defend my stance. `clean-css` tears through it - and my entire dev team is sticking with this package.

I believe `cssnano` and `clean-css` both have their place - why not have dedicated wrappers for each? I'd imagine that like myself, many many many developers are not going to abandon `clean-css` in favor of `cssnano` solely for the execution times I am mentioning. Could `cssnano` improve and optimize these times in the future? Of course so, and probably will - but in the meantime we're stuck with a deprecation warning for a project that is not being maintained - and like all of us - we need it now. I would like to fill that gap and ensure the gulp wrapper is in a good place, is maintained, and offers a reassurance that this plugin is worthwhile to download.

As far as my vision, well, it's a small project, but I can tell you in the short term I definitely see keeping the underlying dependency of `clean-css` up to date, modifying the API as needed with those said updates, and overhauling the unit testing suite. I'll be open and responsive to suggestions and try to make this plugin as robust as I can make it, hopefully with the help of others but with preparation to bare this on my own. 

I hope that is enough to convince you with my sincerity here...

**Comment by phated** [comment]
I'm not questioning your sincerity, we are just moving towards a place where people write less plugins. Instead of wrappers that are suboptimal or out of date, people can use a map stream with the original dependency. That's why I'm looking for reasons to have 2 modules that do relatively the same thing. Thanks for your response.

**Comment by contra** [comment]
@phated That isn't a reason to blacklist. Plugins are still useful, users shouldn't have to string together a bunch of modules and map-stream. The whole point of the plugin is to have uniform standard APIs that work easily with eachother.

**Comment by scniro** [comment]
@phated I do appreciate your angle there, but I would guess with a high degree of confidence that a considerable amount of developers, especially those who are not yet experts with gulp/node ect. will opt for a plugin, and of those some will even base architectural decisions off the availability of a high quality plugin (I've seen it, I've done it, admittedly). Minifying CSS is a hugely popular build step, so with that said I'm quite confident this plugin will continue to serve many developers well.

**Comment by contra** [comment]
@scniro 100% agree with you. Plugins are there for newcomers and people who want assurances that things will work well together.

If you're advanced enough and don't want to use plugins, you also have that choice.

**Comment by phated** [comment]
@contra it is blacklisted because there is another module that minifies CSS, so we need to evaluate if it is more user friendly to recommend using cssnano as the beginner plugin and defer other types of minification, for speed or features, to more advanced use cases.  This is how we've always done the blacklist, I didn't just add this or something, it has been on for quite some time.

**Comment by contra** [comment]
@phated I'm aware of how the blacklist works - I was specifically referring to your comment that people should just use the module directly.

As for this module, It can get more nuanced than picking one - if one is better in some cases, and in other cases the other is better we might have to allow both.

It seems like this one is faster in some cases (as mentioned) - maybe we allow both and put pressure on the underlying modules to fix their issues?

**Comment by scniro** [comment]
@phated it appears to have been blacklisted solely because the plugin was deprecated by the original author as seen here: f850fc23835a80e4c1b6a194186e7f61cc3abddd

Was there an fundamental issue beforehand when both did co-exist?

**Comment by scniro** [comment]
@phated @contra Just as a follow up on this I have merged in the proposal for `gulp-clean-css` to be the recommended gulp plugin per the [clean-css](https://github.com/jakubpawlowicz/clean-css#readme) project repository: 14bb027d1bf6986f7c27671ed03f4d40e08c3bef

Do you have any idea how long a decision on this sort of issue would typically take?

**Comment by scniro** [comment]
the deprecation warning for [gulp-minify-css](https://www.npmjs.com/package/gulp-minify-css) has been updated and now points to `gulp-clean-css` instead of `gulp-cssnano`

**Comment by phated** [comment]
@contra is there a reason to not blacklist gulp-cssnano?  It seems like clean-css is just better?

**Comment by phated** [comment]
I'd be okay blacklisting gulp-cssnano because it also works in a postcss pipeline and https://github.com/postcss/gulp-postcss is a much better way to do postcss stuff in a pipeline (when using gulp-postcss, it only parses once)

**Comment by scniro** [comment]
@phated @contra any thoughts yet on getting `gulp-clean-css` in the plugin registry?

**Comment by phated** [comment]
@scniro see my comments above.  I think we should blacklist gulp-cssnano if we do this.

**Comment by scniro** [comment]
@phated I saw your comments and I'm pretty indifferent on blacklisting `gulp-cssnano`. I think they both have their place, but that's just my opinion. Who's ready to make the call?

**Comment by scniro** [comment]
@phated @contra gentle bump on this issue - any further thoughts from where we last left off?

**Comment by phated** [comment]
Sorry, we are both pretty busy right now.  I will try to follow up by the end of the week.

**Comment by phated** [comment]
@scniro we are good with removing gulp-clean-css from the blacklist but in parallel, we are going to blacklist the `gulp-cssnano` plugin with the reason "use gulp-clean-css or cssnano with gulp-postcss".  If you'd send a PR with these changes, I will merge it.  Thanks!

**Comment by scniro** [comment]
@phated fr-shizzle! PR sent up f7d9f7bb0e13ca190bd2a2f422508598263f9ebe

**Comment by scniro** [comment]
#212

**Comment by contra** [comment]
Just to be clear about the resolution of this: the recommended way to do css minification is to use gulp-postcss with the cssnano plugin. All other gulp plugins that handle css minification should be deprecated and recommend this.

**Comment by scniro** [comment]
@contra agreed. While we are back at it, I do see some dupes in regards to `clean-css` which all seem orphaned. I found these...

- https://www.npmjs.com/package/gulp-mini-css/
- https://www.npmjs.com/package/gulp-minify-css-mpath/
- https://www.npmjs.com/package/gulp-css-prefix/

There were various which did not use the underlying `clean-css` library - but did minify CSS. If your desire is to identify and consolidate all of the plugins doing the same thing - perhaps we can start with these that are mentioned?

**Comment by glen-84** [comment]
I find this whole backlisting concept very strange. I can understand backlisting a plug-in if it:

1. Is deprecated.
2. Is low quality / doesn't follow Gulp standards.
3. Almost entirely duplicates something that already exists (very similar implementation).

But not all CSS minification libraries are created equally. Some have extra features, some have fewer features with better performance, some are more extensible, etc. -- why limit the options?

Perhaps `blackList` is not the right word. Maybe just `unsupported` or similar.

**Comment by scniro** [comment]
@glen-84 I agree on that, and was my main argument coming into this. The whole `cssnano` vs `clean-css` battle (in my case) and which is better was irrelevant to me, since I have specific needs for `clean-css` that `cssnano` couldn't offer. I think there should be a well maintained gulp plugin for each robust css minification library - since like me, many devs are simply not going to use `cssnano` or vice versa. 

Why don't we identify all the major and well supported CSS minifcation libraries, approve a gulp plugin for it, remove any blatant dupes, and live happily ever after?

**Comment by phated** [comment]
@glen-84 no external plugin is supported, blacklist encompasses many things, including but not limited to: filtered from the gulpjs.com/plugins site, failing a gulp-cli `--verify` check and things that can confuse new users.

**Comment by phated** [comment]
@scniro contra already clarified the decision

**Comment by glen-84** [comment]
@phated I was referring to this text from the README:

`Being blacklisted means we won't offer support for issues concerning the module and we will not recommend that people use it.`

Maybe `unsupported` isn't a great suggestion, but my point was that blacklisting has an unfairly negative connotation. If I was the author of a high quality Gulp plug-in wrapping a high quality CSS minification library, I'd be rather annoyed if it was "blacklisted" just because there are other alternatives.

**Comment by phated** [comment]
@glen-84 due to not having a test suite in place that was originally intended, we don't really go through modules unless they are brought to our attention which has lead to lots of css minification plugins, etc.  We aren't about to go through and add them all to the blacklist, as there are just better things to be spending time on. In the case of gulp-cssnano (this thread), it was brought to our attention, we discussed it and made a decision to recommend it as part of gulp-postcss because parsing the file every time for a postcss plugin is a very bad pattern and if you only need a css minification plugin, then we recommend the one @scniro is working on (again, because it was brought to our attention).

**Comment by simison** [comment]
Do you guys know how [gulp-csso](https://github.com/ben-eb/gulp-csso) fits into all this? Seems to be done by same dude as [gulp-cssnano](https://github.com/ben-eb/gulp-cssnano).

**Comment by glen-84** [comment]
@phated,

I have no problem with you recommending one or two plug-ins, which is a great help to inexperienced developers. However, I disagree with statements like this:

> All other gulp plugins that handle css minification should be deprecated and recommend this.

That just seems extreme. =)

Anyway, I appreciate the work that you guys do!

**Comment by contra** [comment]
@glen-84 If it's more nuanced and there are tradeoffs then this is the discussion about that. If something is a straight up duplicate or wrong it should be blacklisted.

Just to be clear to everyone: this is only the plugin search page. You can put anything you want on npm, we have no control over that - but when users type "css minification" on our plugin page we only want them to see things we think should be there.

Nobody is trying to limit the options, but to have 10+ plugins for css minification with vague names like "minify-css" "css-min" "css-minify" and no nuance is confusing as fuck for end-users and we need to fix that.

**Comment by ben-eb** [comment]
Would like to weigh in on this being author of a few CSS minification plugins.

I have been developing cssnano for coming up to a year now, and for me it was an alternative to other minification engines that didn't do what I wanted them to do. I know that there are some issues with large CSS files and deduplicating rules/declarations, and I'm trying to fix this performance issue and provide better merging as a result.

But I have no problem with other competing engines, in fact there are cases in which clean-css outperforms cssnano and vice versa. Perhaps the module will get to a stage where it is the best in every area, don't know for sure. I work on this in my free time, and only have one other main collaborator (@TrySound), so you might appreciate that any big tasks aren't completed with the swiftness. We are also trying to improve PostCSS itself so that we can do cssnano 4 in a better way.

So I think it is right to allow other engines as they fit different use cases.

Blacklisting gulp-cssnano makes sense to me as well, would prefer that people used gulp-postcss especially when loading other PostCSS plugins. I can deprecate this module, the reason that it has stuck around for a while was early plugin collision issues (these go away with a new parse/stringify step), which I *believe* have all been ironed out but I can't say for sure. I think that we'll know for sure once v4 lands, by which time I'll be happy to 'officially' deprecate it.

gulp-csso has been around since the early days of gulp, before cssnano existed. It wraps https://www.npmjs.com/package/csso which is a decent module for minification although it had no maintainers for a while, this partly inspired me to create cssnano, but is now being maintained. I think most people should be looking to use one of those three engines (cssnano, clean-css or csso), but there are alternatives too.

You can have a look at these modules for further details;

* https://github.com/ben-eb/css-minifiers - unified API for all minifiers
* https://github.com/ben-eb/css-minifier-tests - large test suite comparing compression results across engines (~100 tests)

**Comment by phated** [comment]
@ben-eb you rock!  Thanks for all the info!

**Comment by ben-eb** [comment]
Thanks, no problem. :smiley:
